shoes-core/lib/shoes/http_wrapper.rb

```diff
@@ -0,0 +1,55 @@
+require 'uri'
+require 'net/http'
+
+# Wrapper for Net::HTTP to provide us to the functionality that we need which
+# isn't easily available from any particular gem out there :(
+#
+# Specific things this class does that we need include:
+#
+#   1) Reading in chunks to trigger progress
+#   2) Allow for graceful redirection
+#
+# Most of the gems out there do 2) but not 1), or do both but are untested on
+# JRuby.
+#
+class Shoes
+  class HttpWrapper
+    def execute(url, meth, body, headers = {}, redirects_left = 5, &blk)
+      uri = URI(url)
+      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") do |http|
+        request = build_request(uri, meth, body, headers)
+
+        http.request(request) do |response|
+          case response
+          when Net::HTTPRedirection
+            ensure_can_redirect(response, redirects_left)
+            next_location = URI(response["Location"])
+            execute(next_location, "GET", nil, headers, redirects_left - 1, &blk)
+          when Net::HTTPSuccess
+            response.read_body do |chunk|
+              yield response, chunk
+            end
```

Comment:
Yet again, great minds think alike :)

I originally had them in separate methods, but ran into issues with some of the earlier control-flow at the time (I had redirects happening via loop rather than recursion at the time). I see how to move them back now, so I'll do that.